url7f43a4752f_shalynnho_pokemonfactory_tgz-pJ8-Networking_ServerJ8/src/agent/CameraAgent.java

```diff
@@ -0,0 +1,554 @@
+					}
+					// }
+				}
+			}
+		}
+
+		synchronized (nests) {
+			for (MyNest n : nests) {
+				// print("nest "+n.nest.currentPartType+ " state is "+n.state);
+				if (n.state == NestStatus.PHOTOGRAPHED || n.state == NestStatus.RE_PHOTOGRAPHED_ONCE
+						|| n.state == NestStatus.RE_PHOTOGRAPHED_TWICE) {
+					tellPartsRobot(n);
+					return true;
+				}
+			}
+		}
+
+		if (mk != null) {
+			if (mk.ks == KitStatus.NOT_READY) {
+				takePictureOfKit(mk);
+				return true;
+			} else if (mk.ks == KitStatus.DONE) {
+				tellKitRobot(mk);
+				return true;
+			} else if (mk.ks == KitStatus.FAILED) {
+				tellKitRobotFailure(mk);
+				return true;
+			}
+
+		}
+
+		return false;
+	}
+
+	/*
+	 * Actions
+	 */
+	private void tellKitRobot(MyKit kit) {
+		mk = null;
+		kitRobot.msgKitPassedInspection();
+		stateChanged();
+	}
+
+	private void tellKitRobotFailure(MyKit kit) {
+		mk = null;
+		kitRobot.msgKitFailedInspection();
+		stateChanged();
+	}
+
+	private void takePictureOfKit(MyKit kit) {
+		mk.ks = KitStatus.PICTURE_BEING_TAKEN;
+		if (guiCamera != null) {
+			guiCamera.takeKitPhoto(kit.kit.kitGraphics);
+			print("Blocking");
+			try {
+				animation.acquire();
+			} catch (InterruptedException e) {
+				// TODO Auto-generated catch block
+				e.printStackTrace();
+			}
+			print("Got permit");
+		}
+		stateChanged();
+	}
+
+	private void tellPartsRobot(final MyNest n) {
+		List<Part> goodParts = new ArrayList<Part>();
+
+		// Parts missing - Non Norm
+		// if (n.numFilledSnapshot < n.nest.full) {
+		/**
+		 * 1) Nothing, just wait a little longer.This may be the first time parts are coming down the lane and your
+		 * estimation may be wrong. (1) Just wait a little longer and see if parts have arrived.
+		 ** 
+		 * 2) Lane is jammed. (2) hi-speed (turn up the amplitude of the lanes).
+		 ** 
+		 * 6) Your feeder algorithm for keeping parts in two lanes did not changeover fast enough. (6) adjust the
+		 * parameter of your feeder algorithm for changing parts over from one lane to another.
+		 */
+		// print("Missing Parts in Nest : non-normative");
+		// } else {
+		print("nest " + n.nest.currentPartType + " state is " + n.state);
+		if (n.state == NestStatus.PHOTOGRAPHED) {
+			if (n.numFilledSnapshot < n.nest.full) {
+				print("rephotographing nest");
+				n.state = NestStatus.WAITING_TO_RE_PHOTOGRAPH;
+				timer.schedule(new TimerTask() {
+					@Override
+					public void run() {
+						n.state = NestStatus.READY_TO_RE_PHOTOGRAPH;
+					}
+				}, CalculateTimerTime(n));
+			} else {
+				for (MyPart part : n.nest.currentParts) {
+					if (part.part.isGood) {
+						goodParts.add(part.part);
+					}
+				}
+				print("Good parts count: " + goodParts.size());
+
+				// No Good Parts found - non norm
+				if (goodParts.size() == 0) {
+					// print("No Good Parts Found : non-normative");
+					n.nest.msgPurgeSelf();
+				} else {
+					partRobot.msgHereAreGoodParts(n.nest, goodParts);
+				}
+				n.state = NestStatus.NOT_READY;
+			}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 377:393, 406:422, 435:451.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.